Add RT700 basic environment support

lucien-nxp opened the pull request:
support gpio/uart function

lucien-nxp committed:
soc: nxp: imxrt: imxrt7xx: add rt7xx soc files

add rt7xx files related to soc
support basic clock enablement
add common/Kconfig.xspi_xip file

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

lucien-nxp committed:
dts: arm: nxp: add RT7xx dts files

add RT7xx dts files
add iocon/gpio/flexcomm/clock instances in dts

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

lucien-nxp committed:
drivers: clock_control: update mcux_lpc_syscon_clock.c drivers

add more flexcomm instances clock support to adapt
rt700 instances number

add xspi clock support

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

lucien-nxp committed:
drivers: pinctrl: update pinctrl_lpc_iocon.c driver

Define iocon array to store iocon base address
add index parameter support to support multi iocon instances

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

lucien-nxp committed:
drivers: gpio: update gpio_mcux.c driver

update gpio driver to adapt rt7xx gpio model:
1. There is no PORT_Type on RT7xx,so set PORT_Type as void
2. Add port_no parameter in gpio_mcux_config to adapt IOPCTL driver
3. Add gpio-port-offest parameter in blinding, it will help map the
   relation between index n and gpio port when some soc have domain
   access attribution.
3. Splite gpio_mcux_configure function into two functions(
   gpio_mcux_iopctl_configure and gpio_mcux_port_configure)
   according to CONFIG_PINCTRL_NXP_IOCON macro
4. Add code to adapt RT700 GPIO attribute configuration

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

lucien-nxp committed:
boards: nxp: add mimxrt700_evk board

add files related to mimxrt700_evk board
add gpio/uart function support on board

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

lucien-nxp committed:
manifest: update hal_nxp to newest node on main branch

track rt700 sdk files update node

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

danieldegrasse commented on soc/nxp/imxrt/imxrt7xx/Kconfig:
```diff
@@ -0,0 +1,86 @@
+		if $(DT_FLASH_PARENT_IS_XSPI)
+
+config FLASH_SIZE
+	default $(dt_node_int_prop_int,$(DT_CHOSEN_FLASH_NODE),size,Kb) \
```
We should still have the fallthrough case for `CONFIG_FLASH_SIZE` to handle situations where the parent is not XSPI, right? Or does another default definition for this Kconfig come into play?

We should also consider putting these KConfigs into a common file like was done for FlexSPI, since XSPI will likely show up on more SOCs.

danieldegrasse commented on soc/nxp/imxrt/imxrt7xx/Kconfig:
```diff
@@ -0,0 +1,86 @@
+	default y
+	depends on DT_HAS_NXP_LP_FLEXCOMM_ENABLED
+
+config HAS_GLIKEY
```
This seems like it should be placed elsewhere- maybe in `modules/Kconfig.mcux`?

danieldegrasse commented on soc/nxp/imxrt/imxrt7xx/cm33/flash_clock_setup.c:
```diff
@@ -0,0 +1,164 @@
+#include <fsl_cache.h>
+#endif
+
+void BOARD_EnableXspiCache(CACHE64_CTRL_Type *cache)
```
```suggestion
void board_enable_xspi_cache(CACHE64_CTRL_Type *cache)
```
Can we follow Zephyr naming style here?

Also, these should be renamed to not include `board`, since they are at the SOC level.

danieldegrasse commented on soc/nxp/imxrt/imxrt7xx/cm33/soc.c:
```diff
@@ -0,0 +1,186 @@
+ *                 should runs in RAM and move the XSPI clock source to an stable clock to
+ *                 avoid instruction/data fetch issue during clock updating.
+ *END**************************************************************************/
+__attribute__((weak)) void BOARD_XspiClockSafeConfig(void)
```
These are weak, but it seems like we provide strong definitions at the SOC level- should these definitions be at the board level, so users can override them?

lucien-nxp replied:
> These are weak, but it seems like we provide strong definitions at the SOC level- should these definitions be at the board level, so users can override them?

These code like RT500/600 flexspi_clock_safe_config/flexspi_setup_clock function under soc folder, so far the name prefix are just copied from the SDK . Should I modify like RT500/600, or I need to migrate them to board.c?

danieldegrasse replied:
I think the approach we are using now is ok, where the SOC level flexspi function simply sets a mux and a divider.

danieldegrasse commented on soc/nxp/imxrt/imxrt7xx/cm33/soc.c:
```diff
@@ -0,0 +1,186 @@
+{
+}
+
+void z_arm_platform_init(void)
```
```suggestion
void soc_reset_hook(void)
```

See #62595

danieldegrasse commented on soc/nxp/imxrt/soc.yml:
```diff
@@ -38,6 +38,12 @@ family:
+    socs:
+    - name: mimxrt798s
+      cpuclusters:
+      - name: cpu0
```
Same question here- I think we should name these CPU clusters differently given the presence of the HIFI DSP

iuliana-prodan replied:
I agree with @danieldegrasse. 
We also have, on i.MXRT700, 2 DSPs: HiFi4 and HiFi1. We are going to name the two cores simply: hifi1 and hifi4.

nordicjm replied:
This is wrong, cpucluster is the type, see above, so cm33 is a cpu cluster, cpu0/cpu1 is not a cluster

lucien-nxp replied:
Updated to cm33_cpu0/cm33_cpu1. Thank you.

danieldegrasse commented on drivers/pinctrl/pinctrl_lpc_iocon.c:
```diff
@@ -11,22 +11,46 @@
 	for (uint8_t i = 0; i < pin_cnt; i++) {
 		uint32_t pin_mux = pins[i];
+#if defined(CONFIG_SOC_SERIES_IMXRT7XX) && defined(CONFIG_SOC_FAMILY_NXP_IMXRT)
+		uint8_t port_no = PORT_NO(pin_mux);
```
Why can't we use an offset here? The IOPCTL pin control registers appear continuous, just like on the RT600 and RT500

lucien-nxp replied:
Can't use only offset parameter to cover RT700 iopctl model because there are three iopctl instances on the RT700, and the addresses of each instance are not consecutive. The pinctrl header model has been changed to be more compatible with the SDK IOPCTL API function.

danieldegrasse replied:
I'd prefer to just transition the entire driver to use this macro definition if possible.

lucien-nxp replied:
Updated. Thank you.

danieldegrasse commented on drivers/gpio/gpio_mcux.c:
```diff
@@ -77,14 +114,44 @@ static int gpio_mcux_configure(const struct device *dev,
 		return -ENOTSUP;
 	}
 
+#if defined(CONFIG_SOC_SERIES_IMXRT7XX) && defined(CONFIG_SOC_FAMILY_NXP_IMXRT)
```
We should make this dependent on `CONFIG_PINCTRL_NXP_IOCON`, rather than the SOC itself

lucien-nxp replied:
However, I must modify RT700 pinctrl model to adapt RT700 iopctl model. So do you mean I can create a new driver like pinctrl_lpc_iocon_1.c to adapt RT700 new model?

danieldegrasse replied:
The IOCON specific features in this driver still need to be made dependent on `CONFIG_PINCTRL_NXP_IOCON`

danieldegrasse commented on drivers/gpio/gpio_mcux.c:
```diff
@@ -32,19 +42,42 @@ struct gpio_mcux_data {
 	GPIO_Type *gpio_base = config->gpio_base;
+
+#if defined(CONFIG_SOC_SERIES_IMXRT7XX) && defined(CONFIG_SOC_FAMILY_NXP_IMXRT)
+	uint32_t port_no = GPIO_GetInstance(gpio_base);
```
why do we need to calculate this at runtime? Can't we store the port number in the device config structure?

danieldegrasse commented on boards/nxp/mimxrt700_evk/doc/index.rst:
```diff
@@ -0,0 +1,192 @@
+high-performance numerical tasks such as audio and image processing and supports both fixed-point and
+floating-point operations.
+
+Note: Due to board isn't launched, the picture is blank
```
Maybe just don't include an image?  I'm not aware of a requirement that we actually have one.

lucien-nxp replied:
Only release RT700 soc, EVK board don't launch currently. We can't show the board picture currently.

danieldegrasse commented on boards/nxp/mimxrt700_evk/mimxrt700_evk_mimxrt798s_cpu0.dts:
```diff
@@ -0,0 +1,109 @@
+&xspi0 {
+	status = "okay";
+
+	MX25UW51345GXDI00: MX25UW51345GXDI00@0 {
```
Please make this lowercase

danieldegrasse commented on boards/nxp/mimxrt700_evk/doc/index.rst:
```diff
@@ -0,0 +1,192 @@
+For more information about the MIMXRT798 SoC and MIMXRT700-EVK board, see
+these references:
+
+Note: There are currently no available links to provide.
```
I think we can remove notes like these

danieldegrasse commented on boards/nxp/mimxrt700_evk/mimxrt700_evk_mimxrt798s_cpu1_defconfig:
```diff
@@ -0,0 +1,19 @@
+
+# Enable TrustZone-M
+CONFIG_TRUSTED_EXECUTION_SECURE=y
+CONFIG_NXP_IMXRT_BOOT_HEADER=n
```
Missing newline

danieldegrasse commented on boards/nxp/mimxrt700_evk/mimxrt700_evk_mimxrt798s_cpu1_defconfig:
```diff
@@ -0,0 +1,19 @@
+CONFIG_SERIAL=y
+CONFIG_UART_INTERRUPT_DRIVEN=y
+CONFIG_GPIO=y
+CONFIG_PINCTRL=y
```
Just so you know, there is an active effort to remove `CONFIG_PINCTRL` at the board level: #79065. If that PR merges first, this one should be updated.

nordicjm commented on boards/nxp/mimxrt700_evk/Kconfig.defconfig:
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: Apache-2.0
+
+if BOARD_MIMXRT700_EVK
+
```
delete

nordicjm commented on boards/nxp/mimxrt700_evk/doc/mimxrt700_evk.webp:
this image is the concept of a plan for a board, I think

lucien-nxp replied:
Sorry, board won't be launched now, I can't release the board picture.

nordicjm commented on boards/nxp/mimxrt700_evk/mimxrt700_evk-pinctrl.dtsi:
```diff
@@ -0,0 +1,38 @@
+			slew-rate = "normal";
+			drive-strength = "normal";
+		};
+
```
```suggestion
```

nordicjm commented on boards/nxp/mimxrt700_evk/mimxrt700_evk_mimxrt798s_cpu0.dts:
```diff
@@ -0,0 +1,109 @@
+		};
+	};
+
+		gpio_keys {
```
-1x indent

nordicjm commented on boards/nxp/mimxrt700_evk/mimxrt700_evk_mimxrt798s_cpu0.dts:
```diff
@@ -0,0 +1,109 @@
+		status = "okay";
+		jedec-id = [c2 81 3a];
+		erase-block-size = <4096>;
+		write-block-size = <256>;
```
you are telling me this flash chip can only write 256 byte blocks, nothing less than that?

butok replied:
Should write-block-size be 1 or 2 ?

nordicjm commented on boards/nxp/mimxrt700_evk/mimxrt700_evk_mimxrt798s_cpu0_defconfig:
```diff
@@ -0,0 +1,18 @@
+CONFIG_SERIAL=y
+CONFIG_UART_INTERRUPT_DRIVEN=y
+CONFIG_GPIO=y
+CONFIG_PINCTRL=y
```
this should not be here

nordicjm commented on soc/nxp/imxrt/Kconfig:
```diff
@@ -59,6 +68,21 @@ config FLEXSPI_CONFIG_BLOCK_OFFSET
+
+if BOOT_XSPI_NOR
+
+config XSPI_CONFIG_BLOCK_OFFSET
+	hex "XSPI config block offset"
+	default 0x0
+	help
+	  XSPI configuration block consists of parameters regarding specific
+	  flash devices including read command sequence, quad mode enablement
+	  sequence (optional), etc. The boot ROM expects XSPI configuration
+	  parameter to be presented in serial nor flash.
```
why is this not in dts?

danieldegrasse replied:
Does this make sense for DTS? It is a software setting for the offset of the XSPI configuration block, which will then be used by the bootrom to configure the XSPI to interface with an external flash device and execute from it

lucien-nxp replied:
There is no sense for dts. However, we need to use this kconfig to control config block offset in linker files to ensure the offset of the original block.

decsny replied:
@nordicjm Devicetree, as per spec, is meant to be consumed by client program. In this case you can think of the zephyr OS as client program. No software in the Zephyr OS will interact with or consume this config block. It is used by the chip boot ROM, and we only need this offset so we can link it to the proper place in the image in order for the chip to boot. I.e., this config block will be consumed even *before* the zephyr reset handler like `z_arm_reset` runs. I don't think it needs to be in DTS.

nordicjm replied:
I guess it falls between the cracks. Will ask for @gmarull's opinion

nordicjm commented on soc/nxp/imxrt/imxrt7xx/Kconfig.soc:
```diff
@@ -0,0 +1,50 @@
+config SOC_SERIES
+	default "imxrt7xx" if SOC_SERIES_IMXRT7XX
+
+config SOC
+	default "MIMXRT798S" if SOC_MIMXRT798S
```
lowercase

butok commented on boards/nxp/mimxrt700_evk/mimxrt700_evk_mimxrt798s_cpu0.dts:
```diff
@@ -0,0 +1,109 @@
+	status = "okay";
+
+	MX25UW51345GXDI00: MX25UW51345GXDI00@0 {
+		compatible = "nxp,imx-xspi-mx25um51345g";
```
Can it be merged with the nxp,imx-flexspi-mx25um51345g driver?

lucien-nxp replied:
flexspi and xspi are two different spi IP. XSPI is a new IP occurred on nxp soc.

dbaluta commented on dts/arm/nxp/nxp_rt7xx_ns.dtsi:
```diff
@@ -0,0 +1,26 @@
+};
+
+#if defined SOC_MIMXRT798S_CPU0
+#include "nxp_rt7xx_cpu0.dtsi"
```
cpu0 doesn't really tell us much. Can you be more specific.

lucien-nxp replied:
cpu0 is used in RT700 to refer to cm33 cpu0, so I think there is no need to update.

nordicjm replied:
What about `CM33_CPU0` and `cm33_cpu0`?

dbaluta commented on drivers/gpio/gpio_mcux.c:
```diff
@@ -17,6 +17,16 @@
 
 #include <zephyr/drivers/gpio/gpio_utils.h>
 
+#if defined(CONFIG_SOC_SERIES_IMXRT7XX) && defined(CONFIG_SOC_FAMILY_NXP_IMXRT)
```
Please improve the commit message. Update GPIO driver doesn't really tell us much. Please offer a short description of what is the GPIO driver model on imxrt7xx.

dbaluta commented on drivers/gpio/gpio_mcux.c:
```diff
@@ -32,19 +42,42 @@ struct gpio_mcux_data {
+#if defined(CONFIG_SOC_SERIES_IMXRT7XX) && defined(CONFIG_SOC_FAMILY_NXP_IMXRT)
+static uint32_t GPIO_GetInstance(GPIO_Type *base)
+{
+
```
Please notice the coding style of gpio_mcux.c file. Please dont use CamelCase.

so GPIO_GetInstance should go into gpio_get_instance. 

But anyhow, GPIO_GetInstance looks like a function that could come from nxp_hal.

butok replied:
> so GPIO_GetInstance should go into gpio_get_instance.

It is not possible. This is the SDK driver API.

danieldegrasse replied:
In this case it seems we are defining the function within the `gpio_mcux.c` file, right? I'm aware this function is often available within the SDK driver API, but here it does not seem like we are using the SDK API.

lucien-nxp replied:
Updated.

DerekSnell reviewed (changes requested):
@lucien-nxp , thanks for contributing this.

Please consider these changes.  Thank you

lucien-nxp commented:
@hakehuang , sorry I forget to remind you help test. Could you help test basic cases on RT700 board? I will update some content according to the comment above.

hakehuang commented:
@lucien-nxp I meet below problem when flash the image to mimrt700_evk with linkserver see below, it reports
Ed:05: File 'zephyr.bin' load failure: Eg(10). Attempt to load into missing flash area.
```
ubuntu@zephyr-test-06:~/zephyrproject/zephyr$ /usr/local/LinkServer/LinkServer flash --probe QMZ2QFRFA5ZVX -o /device/memory/1/flash-driver=MIMXRT700_XSPI0_Octal.cfx -o /device/memory/1/location=0x38000000 MIMXRT798S:MIMXRT700-EVK load --addr 939524096 /home/ubuntu/zephyrproject/zephyr/twister-out/mimxrt700_evk_mimxrt798s_cpu0/samples/hello_world/sample.basic.helloworld/zephyr/zephyr.bin
INFO: Exact match for MIMXRT798S:MIMXRT700-EVK found
INFO: Selected device MIMXRT798S:MIMXRT700-EVK
INFO: Getting available probes
INFO: Selecting probe by serial substring
INFO: Selected probe #1 QMZ2QFRFA5ZVX (MCU-LINK on-board (r1E6) CMSIS-DAP V3.148)
INFO: MCU-Link firmware update `check`: local firmware [3.148] is the same as the version running on the selected probe ([QMZ2QFRFA5ZVX] [MCU-LINK on-board (r1E6) CMSIS-DAP V3.148])
Firmware update `check`: not required - the update can be performed using `forced` mode
Ns: MCUXpresso IDE RedlinkMulti Driver v24.9 (Sep 19 2024 19:39:44 - crt_emu_cm_redlink build 802)
Pc: (  0) Reading remote configuration
Nc: Found generic directory XML file in /tmp/tmpui7pwq5f/crt_directory.xml
Pc: (  5) Remote configuration complete
Nc: Reconnected to existing LinkServer process.
Wc: ============= SCRIPT: RT700_connect.scp =============
Wc: RT700 Connect Script
Wc: ============= END SCRIPT ============================
Nc: Probe Firmware: MCU-LINK on-board (r1E6) CMSIS-DAP V3.148 (NXP Semiconductors)
Nc: Serial Number:  QMZ2QFRFA5ZVX
Nc: VID:PID:  1FC9:0143
Nc: USB Path: 0001:0002:00
Nc: Using memory from core 0 after searching for a good core
Pc: ( 30) Emulator Connected
Nc: processor is in secure mode
Pc: ( 40) Debug Halt
Pc: ( 50) CPU ID
Nc: debug interface type      = CoreSight DP (DAP DP ID 6BA02477) over SWD TAP 0
Nc: processor type            = Cortex-M33 (CPU ID 00000D21) on DAP AP 0
Nc: number of h/w breakpoints = 8
Nc: number of flash patches   = 0
Nc: number of h/w watchpoints = 4
Nc: Probe(0): Connected&Reset. DpID: 6BA02477. CpuID: 00000D21. Info: <None>
Nc: Debug protocol: SWD. RTCK: Disabled. Vector catch: Disabled.
Ns: Content of CoreSight Debug ROM(s):
Nc: RBASE E00FF000: CID B105100D PID 04000BB4C9 ROM (type 0x1)
Nc: ROM 1 E000E000: CID B105900D PID 04000BBD21 CSt ARM ARMv8-M type 0x0 Misc - Undefined
Nc: ROM 1 E0001000: CID B105900D PID 04000BBD21 CSt ARM DWTv2 type 0x0 Misc - Undefined
Nc: ROM 1 E0002000: CID B105900D PID 04000BBD21 CSt ARM FPBv2 type 0x0 Misc - Undefined
Nc: ROM 1 E0000000: CID B105900D PID 04000BBD21 CSt ARM ITMv2 type 0x43 Trace Source - Bus
Nc: ROM 1 E0041000: CID B105900D PID 04002BBD21 CSt ARM ETMv4.0 type 0x13 Trace Source - Core
Nc: ROM 1 E0042000: CID B105900D PID 04000BBD21 CSt ARM CTIv2 type 0x14 Debug Control - Trigger, e.g. ECT
Nc: CM33 Rev. 1.0  XCACHE
Nc: NXP: MIMXRT798S
Nc: DAP stride is 1024 bytes (256 words)
Nc: Inspected v.2 External Flash Device on SPI MIMXRT700_XSPI0_Octal.cfx
Nc: Image 'iMXRT700_XSPI0_Octal Flash Sep 19 2024 18:21:55'
Nc: Inspected v.2 External Flash Device on SPI MIMXRT700_XSPI0_Octal.cfx
Nc: Image 'iMXRT700_XSPI0_Octal Flash Sep 19 2024 18:21:55'
Nc: Inspected v.2 External Flash Device on SPI MIMXRT700_XSPI0_Octal_S.cfx
Nc: Image 'iMXRT700_XSPI0_Octal Flash Sep 19 2024 18:21:46'
Wc: flash cache - new flash area 'memoryInstance_5' overlaps existing flash area 'memoryInstance_1' - ignored
Nc: Inspected v.2 External Flash Device on SPI MIMXRT700_XSPI1_Octal.cfx
Nc: Image 'iMXRT700_XSPI1_Octal Flash Sep 19 2024 18:22:05'
Nc: Inspected v.2 External Flash Device on SPI MIMXRT700_XSPI1_Octal_S.cfx
Nc: Image 'iMXRT700_XSPI1_Octal Flash Sep 19 2024 18:21:21'
Pc: ( 65) Chip Setup Complete
Pc: ( 70) License Check Complete
Nt: Loading 'zephyr.bin' Binary 0x38000000 len 0xA1F2
Nc: Opening flash driver MIMXRT700_XSPI0_Octal.cfx
Nc: VECTRESET requested, but not supported on ARMv8-M CPUs. Using SOFTRESET instead.
Nc: Using SOFT reset to run the flash driver
Nc: Flash device supported (64MB = 16384*4K at 0x28000000)
Nc: Closing flash driver MIMXRT700_XSPI0_Octal.cfx
Pb: (100) Writing Flash ended with an error.
Ed:05: File 'zephyr.bin' load failure: Eg(10). Attempt to load into missing flash area.
Pc: (100) Target Operation Failed
CRITICAL: Critical error
ERRMSG: Exception: Flash operation exited with code 1
```

lucien-nxp commented:
> @lucien-nxp I meet below problem when flash the image to mimrt700_evk with linkserver see below, it reports Ed:05: File 'zephyr.bin' load failure: Eg(10). Attempt to load into missing flash area.
> 
> ```
> ubuntu@zephyr-test-06:~/zephyrproject/zephyr$ /usr/local/LinkServer/LinkServer flash --probe QMZ2QFRFA5ZVX -o /device/memory/1/flash-driver=MIMXRT700_XSPI0_Octal.cfx -o /device/memory/1/location=0x38000000 MIMXRT798S:MIMXRT700-EVK load --addr 939524096 /home/ubuntu/zephyrproject/zephyr/twister-out/mimxrt700_evk_mimxrt798s_cpu0/samples/hello_world/sample.basic.helloworld/zephyr/zephyr.bin
> INFO: Exact match for MIMXRT798S:MIMXRT700-EVK found
> INFO: Selected device MIMXRT798S:MIMXRT700-EVK
> INFO: Getting available probes
> INFO: Selecting probe by serial substring
> INFO: Selected probe #1 QMZ2QFRFA5ZVX (MCU-LINK on-board (r1E6) CMSIS-DAP V3.148)
> INFO: MCU-Link firmware update `check`: local firmware [3.148] is the same as the version running on the selected probe ([QMZ2QFRFA5ZVX] [MCU-LINK on-board (r1E6) CMSIS-DAP V3.148])
> Firmware update `check`: not required - the update can be performed using `forced` mode
> Ns: MCUXpresso IDE RedlinkMulti Driver v24.9 (Sep 19 2024 19:39:44 - crt_emu_cm_redlink build 802)
> Pc: (  0) Reading remote configuration
> Nc: Found generic directory XML file in /tmp/tmpui7pwq5f/crt_directory.xml
> Pc: (  5) Remote configuration complete
> Nc: Reconnected to existing LinkServer process.
> Wc: ============= SCRIPT: RT700_connect.scp =============
> Wc: RT700 Connect Script
> Wc: ============= END SCRIPT ============================
> Nc: Probe Firmware: MCU-LINK on-board (r1E6) CMSIS-DAP V3.148 (NXP Semiconductors)
> Nc: Serial Number:  QMZ2QFRFA5ZVX
> Nc: VID:PID:  1FC9:0143
> Nc: USB Path: 0001:0002:00
> Nc: Using memory from core 0 after searching for a good core
> Pc: ( 30) Emulator Connected
> Nc: processor is in secure mode
> Pc: ( 40) Debug Halt
> Pc: ( 50) CPU ID
> Nc: debug interface type      = CoreSight DP (DAP DP ID 6BA02477) over SWD TAP 0
> Nc: processor type            = Cortex-M33 (CPU ID 00000D21) on DAP AP 0
> Nc: number of h/w breakpoints = 8
> Nc: number of flash patches   = 0
> Nc: number of h/w watchpoints = 4
> Nc: Probe(0): Connected&Reset. DpID: 6BA02477. CpuID: 00000D21. Info: <None>
> Nc: Debug protocol: SWD. RTCK: Disabled. Vector catch: Disabled.
> Ns: Content of CoreSight Debug ROM(s):
> Nc: RBASE E00FF000: CID B105100D PID 04000BB4C9 ROM (type 0x1)
> Nc: ROM 1 E000E000: CID B105900D PID 04000BBD21 CSt ARM ARMv8-M type 0x0 Misc - Undefined
> Nc: ROM 1 E0001000: CID B105900D PID 04000BBD21 CSt ARM DWTv2 type 0x0 Misc - Undefined
> Nc: ROM 1 E0002000: CID B105900D PID 04000BBD21 CSt ARM FPBv2 type 0x0 Misc - Undefined
> Nc: ROM 1 E0000000: CID B105900D PID 04000BBD21 CSt ARM ITMv2 type 0x43 Trace Source - Bus
> Nc: ROM 1 E0041000: CID B105900D PID 04002BBD21 CSt ARM ETMv4.0 type 0x13 Trace Source - Core
> Nc: ROM 1 E0042000: CID B105900D PID 04000BBD21 CSt ARM CTIv2 type 0x14 Debug Control - Trigger, e.g. ECT
> Nc: CM33 Rev. 1.0  XCACHE
> Nc: NXP: MIMXRT798S
> Nc: DAP stride is 1024 bytes (256 words)
> Nc: Inspected v.2 External Flash Device on SPI MIMXRT700_XSPI0_Octal.cfx
> Nc: Image 'iMXRT700_XSPI0_Octal Flash Sep 19 2024 18:21:55'
> Nc: Inspected v.2 External Flash Device on SPI MIMXRT700_XSPI0_Octal.cfx
> Nc: Image 'iMXRT700_XSPI0_Octal Flash Sep 19 2024 18:21:55'
> Nc: Inspected v.2 External Flash Device on SPI MIMXRT700_XSPI0_Octal_S.cfx
> Nc: Image 'iMXRT700_XSPI0_Octal Flash Sep 19 2024 18:21:46'
> Wc: flash cache - new flash area 'memoryInstance_5' overlaps existing flash area 'memoryInstance_1' - ignored
> Nc: Inspected v.2 External Flash Device on SPI MIMXRT700_XSPI1_Octal.cfx
> Nc: Image 'iMXRT700_XSPI1_Octal Flash Sep 19 2024 18:22:05'
> Nc: Inspected v.2 External Flash Device on SPI MIMXRT700_XSPI1_Octal_S.cfx
> Nc: Image 'iMXRT700_XSPI1_Octal Flash Sep 19 2024 18:21:21'
> Pc: ( 65) Chip Setup Complete
> Pc: ( 70) License Check Complete
> Nt: Loading 'zephyr.bin' Binary 0x38000000 len 0xA1F2
> Nc: Opening flash driver MIMXRT700_XSPI0_Octal.cfx
> Nc: VECTRESET requested, but not supported on ARMv8-M CPUs. Using SOFTRESET instead.
> Nc: Using SOFT reset to run the flash driver
> Nc: Flash device supported (64MB = 16384*4K at 0x28000000)
> Nc: Closing flash driver MIMXRT700_XSPI0_Octal.cfx
> Pb: (100) Writing Flash ended with an error.
> Ed:05: File 'zephyr.bin' load failure: Eg(10). Attempt to load into missing flash area.
> Pc: (100) Target Operation Failed
> CRITICAL: Critical error
> ERRMSG: Exception: Flash operation exited with code 1
> ```

I have deleted the wrong address in board.cmake, please help retry and test.  I will update code according comment above on these several days.

hakehuang commented:
> I have deleted the wrong address in board.cmake, please help retry and test. I will update code according comment above on these several days.

@Lucien-Zhao , now I can flash, but there is no console output for hello_world or any applications, and debug found the entry is error. please check.

lucien-nxp commented:
> > I have deleted the wrong address in board.cmake, please help retry and test. I will update code according comment above on these several days.
> 
> @Lucien-Zhao , now I can flash, but there is no console output for hello_world or any applications, and debug found the entry is error. please check.

Hi Hake,
I suspect your boards is in wrong boot mode. Please check sw10 boot mode whether is on 01 status. Thank you.

decsny commented on boards/nxp/mimxrt700_evk/board.c:
```diff
@@ -0,0 +1,263 @@
+	GlikeyClearConfig(GLIKEY2);
+}
+
+SYS_INIT(mimxrt700_evk_init, PRE_KERNEL_1, CONFIG_BOARD_INIT_PRIORITY);
```
use board hooks

decsny commented on boards/nxp/mimxrt700_evk/board.c:
```diff
@@ -0,0 +1,263 @@
+	RESET_ClearPeripheralReset(kGPIO7_RST_SHIFT_RSTn);
+#endif
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(gpio8), okay)
```
use if() not #if  to increase compilation coverage and reduce conditional code (compiler optimizes away things in if(0))

lucien-nxp replied:
I have tried it, however I meet an issue when changed to not use conditional code. Due to domain, cpu0/1 can't access all the peripheral, if I use  C if as judgement, C complier still check the element under the if statement. So it means that C compiler will report an error don't find IP cpu1 can access(however, cpu0 can't) when I compile a cpu0 image.

lucien-nxp replied:
This seems impossible because the premise for using this is that DT_DRV_COMPAT is unique, but there are too many types of IP in board.c.

decsny commented on boards/nxp/mimxrt700_evk/board.c:
```diff
@@ -0,0 +1,263 @@
+/* Disable the secure check for AHBSC and enable periperhals/sram access for masters */
+static void BOARD_InitAHBSC(void)
+{
+#if defined(CONFIG_SOC_MIMXRT798S_CPU0)
```
can probably use if() here as well

lucien-nxp replied:
if this is changed, the compiler will report an error, CONFIG_SOC_MIMXRT798S_CPU0 cannot be found

decsny commented on boards/nxp/mimxrt700_evk/Kconfig:
file not needed if using board hooks

decsny commented on soc/nxp/imxrt/imxrt7xx/cm33/soc.c:
```diff
@@ -0,0 +1,186 @@
+	CLOCK_AttachClk(kCOMMON_BASE_to_COMMON_VDDN);
+	CLOCK_SetClkDiv(kCLOCK_DivCommonVddnClk, 1U);
+
+#ifdef CONFIG_FLASH_MCUX_XSPI_XIP
```
if()

lucien-nxp replied:
Same like above.

decsny commented on soc/nxp/imxrt/imxrt7xx/cm33/soc.c:
```diff
@@ -0,0 +1,186 @@
+		.coarseTrimEn = true,
+	};
+
+#ifndef CONFIG_IMXRT7XX_CODE_CACHE
```
if()

lucien-nxp replied:
Same like above.

decsny commented on soc/nxp/imxrt/imxrt7xx/cm33/soc.c:
```diff
@@ -0,0 +1,186 @@
+
+static int imxrt7xx_soc_init(void)
+{
+#ifdef CONFIG_SOC_MIMXRT798S_CPU0
```
if() wherever possible for whole PR (sometimes, you have to use #if , but usually not)

lucien-nxp replied:
Same like above.

decsny commented on modules/Kconfig.mcux:
```diff
@@ -96,6 +96,11 @@ config HAS_MCUX_FLEXSPI
 	help
 	  Set if the flexible SPI (FlexSPI) module is present in the SoC.
 
+config HAS_MCUX_XSPI
```
shouldn't be needed, these HAS_MCUX_XXX are legacy, use DT

decsny commented on dts/arm/nxp/nxp_rt7xx_ns.dtsi:
```diff
@@ -0,0 +1,26 @@
+	};
+};
+
+#if defined SOC_MIMXRT798S_CPU0
```
DT is processed before kconfig, this is not going to work, and I'm assuming this file has not been actually tested against any build ? if so, just remove this file if not enabling ns in this PR

decsny commented on dts/arm/nxp/nxp_rt7xx_cpu1.dtsi:
```diff
@@ -0,0 +1,350 @@
+			ranges = <0x0 0x50000000 0x10000000>;
+		};
+
+		xspi2: spi@50411000 {
```
make an xspi binding, even if its just a compatible string for now, use instead of kconfig HAS_MCUX_XSPI

decsny commented on soc/nxp/imxrt/imxrt7xx/Kconfig.defconfig:
```diff
@@ -0,0 +1,31 @@
+config FLASH_SIZE
+	default 65536
+
+config ZTEST_NO_YIELD
```
why is this set here?

lucien-nxp replied:
Deleted.

decsny commented on soc/nxp/imxrt/imxrt7xx/cm33/soc.h:
does this file actually need to exist?

lucien-nxp replied:
Yes, due to cmsis_core_.h  need to include it.

decsny commented on boards/nxp/mimxrt700_evk/board.c:
```diff
@@ -0,0 +1,263 @@
+
+static void BOARD_InitAHBSC(void);
+
+#define SET_UP_FLEXCOMM_CLOCK(x)                                                                   \
```
good idea

lucien-nxp commented on boards/nxp/mimxrt700_evk/mimxrt700_evk_mimxrt798s_cpu0.dts:
```diff
@@ -0,0 +1,109 @@
+};
+
+&flexcomm8{
+	status = "okay";
```
OK, I will add comment here. There's no need to add pinctrl settings for FC8 because GPIO function is default like other RT500/600.

danieldegrasse replied:
We should still set the pin control settings, there is no promise within Zephyr that a prior bootloader or similar didn't change the pin configuration.

decsny commented on soc/nxp/imxrt/imxrt7xx/Kconfig.defconfig:
```diff
@@ -0,0 +1,31 @@
+config NUM_IRQS
+	default 93
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 100000000 if CORTEX_M_SYSTICK
```
this is another case of this is where it is for all our socs

decsny replied:
on a related note, it would be really useful i think if we standardized in zephyr the kernel hw timer being a dt chosen node

lucien-nxp replied:
Correct the value to 237500000

lucien-nxp commented on boards/nxp/mimxrt700_evk/doc/index.rst:
```diff
@@ -0,0 +1,192 @@
+========
+
+Here is an example for the :zephyr:code-sample:`hello_world` application. This example uses the
+:ref:`jlink-debug-host-tools` as default.
```
I have updated to use JLINK as default debug way like other RT500/600 platforms.

lucien-nxp replied:
I have updated to use JLINK as default debug way like other RT500/600 platforms.

lucien-nxp commented on boards/nxp/mimxrt700_evk/doc/mimxrt700_evk.webp:
I remember if I delete picture, doc test can't be tested. So I put on a blank picture here.

nordicjm replied:
@kartben ?

lucien-nxp replied:
I have deleted it and don't describe the picture in index.rst. There is no doc issue reported.

lucien-nxp committed:
dts: bindings: add the first version binding for nxp,xspi IP

add nxp,xspi-device.yaml
add nxp,xspi-mx25um51345g.yaml
add nxp,xspi.yaml

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

nordicjm commented on boards/nxp/mimxrt700_evk/Kconfig.mimxrt700_evk:
```diff
@@ -0,0 +1,7 @@
+# Copyright 2024 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_MIMXRT700_EVK
+	select SOC_MIMXRT798S_CPU0 if BOARD_MIMXRT700_EVK_MIMXRT798S_CPU0
+	select SOC_MIMXRT798S_CPU1 if BOARD_MIMXRT700_EVK_MIMXRT798S_CPU1
```
this is wrong, this is the CM33 CPU cluster, it should be using it as such

danieldegrasse commented on boards/nxp/mimxrt700_evk/mimxrt700_evk_mimxrt798s_cpu0.dts:
```diff
@@ -0,0 +1,109 @@
+		status = "okay";
+		jedec-id = [c2 81 3a];
+		erase-block-size = <4096>;
+		write-block-size = <2>; /* FLASH_MCUX_FLEXSPI_MX25UM51345G_OPI_DTR set */
```
This doesn't really need to be present, since we have no enablement for XSPI (so this property won't affect anything)

lucien-nxp replied:
Yes, due to the same flash reused on RT700, I only copied the content from RT500. So do you mean I need to remove these information?

lucien-nxp replied:
About flexcomm8 set as ok status:  I'm sorry I don't know why I added it then(The brain might be down😀)and I have deleted it. Thank you

lucien-nxp replied:
Deleted.

danieldegrasse commented on boards/nxp/mimxrt700_evk/CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+if(CONFIG_BOARD_MIMXRT700_EVK_MIMXRT798S_CPU0 OR CONFIG_BOARD_MIMXRT700_EVK_MIMXRT798S_CPU1)
+  zephyr_library()
+  zephyr_library_sources(board.c)
+  zephyr_library_include_directories(.)
```
Why does the board directory need to be added to the include search path?

danieldegrasse commented on boards/nxp/mimxrt700_evk/doc/index.rst:
```diff
@@ -0,0 +1,192 @@
+    .. group-tab:: Linkserver
+
+        1. Install the :ref:`linkserver-debug-host-tools` and make sure they are in your search path.
+        2. To update the debug firmware, please follow the instructions on `MIMXRT700-EVK Debug Firmware`
```
To add to this, we have reusable sets of instructions if this board uses an MCULink probe: https://github.com/zephyrproject-rtos/zephyr/blob/372b3a9b8da2967e43676943392daf9091affdb1/boards/nxp/frdm_mcxn947/doc/index.rst?plain=1#L159

danieldegrasse commented on drivers/gpio/gpio_mcux.c:
```diff
@@ -77,14 +95,44 @@ static int gpio_mcux_configure(const struct device *dev,
 		return -ENOTSUP;
 	}
 
+#if defined(CONFIG_SOC_SERIES_IMXRT7XX)
```
Should depend on `CONFIG_PINCTRL_NXP_IOCON`

danieldegrasse commented on drivers/gpio/gpio_mcux.c:
```diff
@@ -261,49 +305,48 @@ static uint32_t get_gpio_icr_irqc_value_from_flags(const struct device *dev,
 {
 	const struct gpio_mcux_config *config = dev->config;
 	GPIO_Type *gpio_base = config->gpio_base;
+#if !(defined(CONFIG_SOC_SERIES_IMXRT7XX))
```
Should depend on `CONFIG_PINCTRL_NXP_IOCON`

danieldegrasse commented on drivers/pinctrl/pinctrl_lpc_iocon.c:
```diff
@@ -11,23 +11,40 @@
 #endif
 
+/* RT7xx requests to clear the reset bit in RESET module before a IOPCTL can be used. */
+#if defined(CONFIG_SOC_SERIES_IMXRT7XX)
```
So right now we are essentially using a different mask definition for the RT700? I think we should probably use some DT property on the iocon node to control if we use this mask definition or not, or we should transition *all* SOCs using this pin controller to use the new macro definition.

@decsny any thoughts here on the usage of DT to control this behavior?

danieldegrasse commented on soc/nxp/common/Kconfig.xspi_xip:
```diff
@@ -0,0 +1,42 @@
+config CODE_DATA_RELOCATION_SRAM
+	default y
+
+config FLASH_MCUX_XSPI_XIP_MEM
```
Where are we using this Kconfig?

lucien-nxp replied:
Copy it from flexspi side.  Currently, I have applied the kconfig into cmakelists under cm33 folder to ensure flash driver executed in RAM region.

danieldegrasse commented on soc/nxp/imxrt/imxrt7xx/cm33/flash_clock_setup.c:
```diff
@@ -0,0 +1,168 @@
+}
+
+/* flexspi_setup_clock run in RAM when XIP. */
+void flexspi_setup_clock(XSPI_Type *base, uint32_t src, uint32_t divider)
```
```suggestion
void xspi_setup_clock(XSPI_Type *base, uint32_t src, uint32_t divider)
```

danieldegrasse commented on soc/nxp/imxrt/imxrt7xx/cm33/flash_clock_setup.h:
```diff
@@ -0,0 +1,18 @@
+#include "MIMXRT798S_cm33_core0.h"
+#endif
+
+void flexspi_clock_safe_config(void);
```
```suggestion
void xspi_clock_safe_config(void);
```

danieldegrasse commented on soc/nxp/imxrt/imxrt7xx/cm33/flash_clock_setup.h:
```diff
@@ -0,0 +1,18 @@
+#endif
+
+void flexspi_clock_safe_config(void);
+void flexspi_setup_clock(XSPI_Type *base, uint32_t src, uint32_t divider);
```
```suggestion
void xspi_setup_clock(XSPI_Type *base, uint32_t src, uint32_t divider);
```

DerekSnell commented on soc/nxp/imxrt/imxrt7xx/cm33/soc.c:
```diff
@@ -0,0 +1,163 @@
+	SystemInit();
+}
+
+void soc_early_init_hook(void)
```
Most of this clock code configured in `soc_early_init_hook()` is board/application specific.  If users want to modify these clocks to match their board, it is not easily done.  Please move all this clock init code to `board_early_init_hook()` in board.c.  Then users can easily modify with their custom board.  Thanks

DerekSnell reviewed (changes requested):
Hi @lucien-nxp ,
Thank you for addressing my other feedback.  Now I just see 1 simple doc change, but an important one.  Thank you

DerekSnell commented on boards/nxp/mimxrt700_evk/doc/index.rst:
```diff
@@ -0,0 +1,200 @@
+Using LinkServer
+----------------
+
+Linkserver is the default runner for this board, and supports the factory
```
```suggestion
The Linkserver runner supports the factory
```
JLink is the default runner in the PR.  We need to correct this.

lucien-nxp replied:
Updated, thank you for reminding.

DerekSnell reviewed (dismissed):
Hi @lucien-nxp ,
Thank you for addressing my other feedback. Now I just see 1 simple doc change, but an important one. Thank you

lucien-nxp commented:
I created new pinctrl driver(iopctl driver) for RT 3 digital platforms, it makes better service the multi iopctl instances on RT700, and keep a same pinctrl model for RT 3 digital platforms. So this PR is currently dependent on my other [PR#81086](https://github.com/zephyrproject-rtos/zephyr/pull/81086).

kartben commented:
@lucien-nxp looks like there was a rebase accident :)

lucien-nxp commented:
> @lucien-nxp looks like there was a rebase accident :)

Due to this PR is dependent on 

> @lucien-nxp looks like there was a rebase accident :)



> @lucien-nxp looks like there was a rebase accident :)

lucien-nxp commented:
> @lucien-nxp looks like there was a rebase accident :)
This PR is dependent on another [#81086](https://github.com/zephyrproject-rtos/zephyr/pull/81086). I will speed another PR merged, then rebase this PR the newest main node.  Currently, only do ci test on this PR. Thank you.

DerekSnell reviewed (commented):
Hi @lucien-nxp ,
Thank you for addressing all my requests.  I am removing my block.  But I will let other reviewers address their requests and the IOPCTL changes.  Thank you

lucien-nxp commented on dts/arm/nxp/nxp_rt7xx_cm33_cpu1.dtsi:
```diff
@@ -0,0 +1,353 @@
+/*
```
I think there is no need to create common dtsi for RT700, domain makes two core access IP instance is different.

lucien-nxp commented:
Hi  @Raymond0225 ,
Now iocon driver won't be used for RT 3 digital platforms. I have created new IOPCTL driver to service RT 3 digital platforms.

lucien-nxp commented:
Hi @hakehuang ， Could you help test all the LPC and RT3 digitals platforms based on this PR?  I have updated iocon driver and pin header file for RT700.  Thank you.

hakehuang commented:
> Hi @hakehuang ， Could you help test all the LPC and RT3 digitals platforms based on this PR? I have updated iocon driver and pin header file for RT700. Thank you.

sure, I kick another round of testing, will feedback once done

lucien-nxp commented:
Hi @danieldegrasse ,
Could you help review the current iocon driver and pinctrl model on RT700?   I have updated them according to the model you provide.

lucien-nxp committed:
samples: userspace: shared_mem: remove mimxrt700_evk cm33 cores

The number of mpu regions that can be configured is less than four cases.
Therefore, only remove this case on cm33 cores, failed log show below:
"num_parts of 4 exceeds maximum allowable partitions (3)"

samples: kernel: mem_protect: userspace: remove mimxrt700_evk cm33 cores

The maximum number of partitions is less than the number of partitions
currently required, causing case to be unable to apply for or delete
excess regions.

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

lucien-nxp commented on soc/nxp/imxrt/imxrt7xx/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+if (CONFIG_SOC_MIMXRT798S_CM33_CPU0 OR CONFIG_SOC_MIMXRT798S_CM33_CPU1)
```
Updated.

lucien-nxp commented on boards/nxp/mimxrt700_evk/doc/index.rst:
```diff
@@ -0,0 +1,199 @@
+family of MCUs.  This board is a focus for NXP's Full Platform Support for
+Zephyr, to better enable the entire RT7xx family.  NXP prioritizes enabling
+this board with new support for Zephyr features.  The ``mimxrt700_evk/mimxrt798s
+/cm33_cpu0`` or  ``mimxrt700_evk/mimxrt798s/cm33_cpu1`` board configuration supports
```
Updated.

lucien-nxp commented on soc/nxp/common/Kconfig.xspi_xip:
```diff
@@ -0,0 +1,42 @@
+	default y
+
+config FLASH_MCUX_XSPI_XIP_MEM
+	string
+	prompt "Xspi drivers memory location"
+	default "RAM"
+	help
+	  Select the location to run the XSPI drivers when using
+	  the flash API.
```
No, it should be string because it will be used in cmake files, need to specifies the name of the region

nordicjm replied:
I'm not understanding this, the region? What region? What possible values can this Kconfig have and what does it do?

lucien-nxp replied:
It should be a string because it will be used in CMakeLists to relocate some [CMakeLists](https://github.com/zephyrproject-rtos/zephyr/pull/79376/files#diff-cccf11441beec6ffb893f5013736668369af7c6cad7da460e58ff8cc0506a3d7:~:text=LOCATION%20%24%7BCONFIG_FLASH_MCUX_XSPI_XIP_MEM%7D_TEXT))

lucien-nxp commented:
> > Hi @hakehuang ， Could you help test all the LPC and RT3 digitals platforms based on this PR? I have updated iocon driver and pin header file for RT700. Thank you.
> 
> sure, I kick another round of testing, will feedback once done
> 
> @lucien-nxp
> 
> op 500 values of testsuite.keyword	Failed
> tests/kernel/mem_protect/mem_protect/kernel.memory_protection	15
> tests/kernel/mem_protect/userspace/kernel.memory_protection.userspace	1
> tests/kernel/threads/thread_stack/kernel.threads.armv8m_mpu_stack_guard	1
> tests/kernel/fatal/exception/kernel.common.stack_protection_armv8m_mpu_stack_guard	2
> samples/userspace/prod_consumer/sample.userspace.prod_consumer	1
> samples/userspace/shared_mem/sample.kernel.memory_protection.shared_mem	1

Hi Hake,
These issue seems is caused by limited MPU region available on cm33 core. I have updated the code and try most of them, and it have been passed on my local.  Could you help re-test?

lucien-nxp commented:
> Hi @danieldegrasse , Could you help review the current iocon driver and pinctrl model on RT700? I have updated them according to the model you provide.

Hi @danieldegrasse , could you help review the current pinctrl model?  Thank you.

danieldegrasse commented on soc/nxp/imxrt/imxrt7xx/Kconfig:
```diff
@@ -0,0 +1,57 @@
+	select HAS_MCUX
+	select HAS_MCUX_SYSCON
+	select HAS_GLIKEY
+	select SOC_RESET_HOOK
```
Will CPU core 1 need the reset hook to run? Just asking since the secondary core would only be booted by the primary, right?

lucien-nxp replied:
Yes, I have removed it for cm33_cpu1.

danieldegrasse commented on dts/bindings/mtd/nxp,imx-xspi-mx25um51345g.yaml:
```diff
@@ -0,0 +1,8 @@
+
+description: NXP XSPI MX25UM51345G
+
+compatible: "nxp,imx-xspi-mx25um51345g"
```
Let's avoid using `imx-` in the naming scheme, just use `nxp,xpi`

lucien-nxp replied:
Updated.

danieldegrasse commented on dts/bindings/spi/nxp,imx-xspi.yaml:
```diff
@@ -0,0 +1,20 @@
+
+description: NXP XSPI controller
+
+compatible: "nxp,imx-xspi"
```
```suggestion
compatible: "nxp,xspi"
```

Avoiding family names in compatibles is best because we may reuse this IP on non-IMX.RT cores

lucien-nxp replied:
Updated.

danieldegrasse commented on drivers/pinctrl/pinctrl_lpc_iocon.c:
Commit message here needs to be updated, `IOPCTL_PinMuxSet` function is not being used

lucien-nxp replied:
Update information in commit message. Thank you.

danieldegrasse commented on drivers/pinctrl/pinctrl_lpc_iocon.c:
```diff
@@ -11,22 +11,40 @@
 #endif
 
+/* IOCON register addresses. */
+static uint32_t *iocon[] = {
```
```suggestion
static uint32_t *volatile iocon[] = {
```
I think the pointers in this array need to be declared as volatile still

lucien-nxp replied:
Updated.

danieldegrasse commented on drivers/pinctrl/pinctrl_lpc_iocon.c:
```diff
@@ -11,22 +11,40 @@
 	for (uint8_t i = 0; i < pin_cnt; i++) {
 		uint32_t pin_mux = pins[i];
 		uint32_t offset = OFFSET(pin_mux);
+		uint32_t index = INDEX(pin_mux);
```
No need to use `uint32_t` here, this can be a `uint8_t` based on mask width, right?

lucien-nxp replied:
Updated.

danieldegrasse commented on drivers/gpio/gpio_mcux.c:
```diff
@@ -32,19 +42,23 @@ struct gpio_mcux_data {
 	const struct gpio_mcux_config *config = dev->config;
 	GPIO_Type *gpio_base = config->gpio_base;
+
+#if defined(CONFIG_PINCTRL_NXP_IOCON)
```
Perhaps rather than doing all this `#ifdef`'ery, could we split the pin control settings into two functions?
- `gpio_mcux_port_configure`
- `gpio_mcux_iopctl_configure`

Then we could simply have one `ifdef` in this function, and call the correct helper

lucien-nxp replied:
Split to two functions. Updated.

danieldegrasse commented on drivers/gpio/gpio_mcux.c:
```diff
@@ -404,47 +445,44 @@ static DEVICE_API(gpio, gpio_mcux_driver_api) = {
-									\
-	static int gpio_mcux_port## n ##_init(const struct device *dev)	\
-	{								\
+#define GPIO_PORT_NUMBER(n) COND_CODE_1(DT_INST_NODE_HAS_PROP(n, gpio_port_offest),	\
```
If possible splitting the formatting changes here (which seem to be from running clang-format) out from the functional changes to the driver would make this easier to review. If not, just be aware that is best practice in the future (not to mix formatting changes and functional ones in the same commit)

lucien-nxp replied:
Split the format changes. Thank you..

danieldegrasse reviewed (commented):
IOCON driver changes generally look good, thank you for this update

lucien-nxp commented on modules/Kconfig.mcux:
```diff
@@ -394,4 +394,9 @@ config NXP_RF_IMU
 	help
 	   RF_IMU adapter is needed for intercore messaging.
 
+config HAS_GLIKEY
```
Currently, we can't describe it in dts, there is no driver supported. I will migrate Kconfig defined under rt7xx folder and modify the kconfig name to GLIKEY_MCUX_GLIKEY.

danieldegrasse replied:
We can still have a devicetree node without a specific driver- for example, I've done this for PLL configurations within SOCs:

https://github.com/zephyrproject-rtos/zephyr/blob/e60f04096cd8d9227c8279ca830e72bc1b5b8b1c/soc/nxp/imxrt/imxrt10xx/soc.c#L45
https://github.com/zephyrproject-rtos/zephyr/blob/e60f04096cd8d9227c8279ca830e72bc1b5b8b1c/dts/bindings/clock/nxp,imx-ccm-fnpll.yaml#L1

lucien-nxp replied:
Here I don't need to get parameter from dts, only call the SDK driver.

lucien-nxp commented on boards/nxp/mimxrt700_evk/doc/index.rst:
```diff
@@ -0,0 +1,199 @@
+.. _mimxrt700_evk:
+
+NXP MIMXRT700-EVK
+##################
```
Thank you for reminding. The board name is mimxrt700_evk, don't like rt500/600 platform.  
Due to board still not released externally, currently I can't update board picture.

lucien-nxp replied:
For Q1 I have update zephyr:board for mimxrt700_evk. Thank you.
For Q2，I can't put picture because the rt700 board don't released externally currently.

kartben replied:
This hasn't been addressed?

lucien-nxp replied:
Yes, RT700 board won't be released externally recently.

kartben replied:
I'm sorry, I don't understand. You commented to indicate you have applied "zephyr:board::" directive but you haven't, so I'm confused? It's ok if you don't want to include an image at the moment, but the board documentation page should be correct/complete nevertheless

lucien-nxp replied:
Now  "zephyr:board::" directive had been added into index.rst file. I'm not clear about your mean on the documentation page. What is correct/complete for board documentation page?

lucien-nxp committed:
drivers: gpio: format gpio_mcux.c

use clang-format to format gpio_mcux file

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

lucien-nxp commented on drivers/gpio/gpio_mcux.c:
```diff
@@ -17,6 +17,16 @@
 
 #include <zephyr/drivers/gpio/gpio_utils.h>
 
+#if defined(CONFIG_SOC_SERIES_IMXRT7XX) && defined(CONFIG_SOC_FAMILY_NXP_IMXRT)
```
Instead of using Kconfig with the NPI name, use the IP name: CONFIG_PINCTRL_NXP_IOCON

lucien-nxp commented on soc/nxp/common/Kconfig.xspi_xip:
```diff
@@ -0,0 +1,42 @@
+# SPDX-License-Identifier: Apache-2.0
+
+DT_CHOSEN_Z_FLASH := zephyr,flash
+DT_COMPAT_XSPI := nxp,imx-xspi
```
updated.

lucien-nxp commented on west.yml:
Updated.

hakehuang commented:
@lucien-nxp , board testing on RT700 works fine now. and RTxxx hello_world works as well..

lucien-nxp commented:
Only rebase to the newest node to resolve the west.yml conflict.

lucien-nxp commented:
Only rebase to the newest node to resolve the west.yml conflict.

lucien-nxp commented:
Only rebase to the newest node to resolve the west.yml conflict.

kartben commented on boards/nxp/mimxrt700_evk/board.yml:
```diff
@@ -0,0 +1,5 @@
+board:
+  name: mimxrt700_evk
```
```suggestion
  name: mimxrt700_evk
  full_name: MIMXRT700-EVK
```

lucien-nxp replied:
Updated. Thank you for pointing out.

kartben commented on boards/nxp/mimxrt700_evk/doc/index.rst:
```diff
@@ -0,0 +1,199 @@
+.. zephyr:board:: mimxrt700_evk
+
+NXP MIMXRT700-EVK
+#################
```
```suggestion
```

lucien-nxp replied:
Deleted.

kartben commented on boards/nxp/mimxrt700_evk/doc/index.rst:
```diff
@@ -0,0 +1,199 @@
+Hardware
+********
+
+- Main Compute Subsystem:
+      Arm Cortex-M33 up to 325 MHz
+      HiFi 4 DSP up to 325 MHz
+      eIQ Neutron NPU up to 325 MHz
+- Sense Compute Subsystem:
+      Arm Cortex-M33 up to 250 MHz
+      HiFi 1 DSP up to 250 MHz
```
Incorrectly formatted sub bullet list

```
- list item

  - sublist item
  - ...
```

Note the blank line needed before starting sublist

lucien-nxp replied:
Updated. I will check later when online doc generated.

kartben reviewed (changes requested):
You may preview your board documentation in CI

https://builds.zephyrproject.io/zephyr/pr/79376/docs/boards/index.html#name=700

https://builds.zephyrproject.io/zephyr/pr/79376/docs/boards/nxp/mimxrt700_evk/doc/index.html

kartben commented on boards/nxp/mimxrt700_evk/doc/index.rst:
```diff
@@ -0,0 +1,192 @@
+Hardware
+********
+
+- Main Compute Subsystem:
+   - Arm Cortex-M33 up to 325 MHz
+   - HiFi 4 DSP up to 325 MHz
+   - eIQ Neutron NPU up to 325 MHz
+- Sense Compute Subsystem:
+   - Arm Cortex-M33 up to 250 MHz
+   - HiFi 1 DSP up to 250 MHz
```
As per my earlier comment this should be:

```suggestion
- Main Compute Subsystem:

  - Arm Cortex-M33 up to 325 MHz
  - HiFi 4 DSP up to 325 MHz
  - eIQ Neutron NPU up to 325 MHz
- Sense Compute Subsystem:

  - Arm Cortex-M33 up to 250 MHz
  - HiFi 1 DSP up to 250 MHz
```

lucien-nxp replied:
Sorry, I will check the generated doc online and update it together. Thank you.

lucien-nxp replied:
May I ask why we need to empty line?  I checked the doc generated, and I think there is no issue on generated doc：
![Uploading image.png…]()

kartben replied:
because what you're seeing here is a "definition list" (see https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#definition-lists). Not how "Main Compute subsytem" and "Sense Compute Subsytem" are highlighted and how they are styled differently as the other items of the same level ("7.5 MB on-chip SRAM", etc.)
Adding the empty line and having correct alignment of the sub-list as per my suggested change ensures this is properly formatted as a bullet/sub-bullet list (see https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#bullet-lists)

lucien-nxp replied:
Thank you, I have updated.

lucien-nxp commented:
Only rebase to the newest commit on main branch and resolve the west.yml conflicts.

decsny commented on soc/nxp/common/Kconfig.xspi_xip:
This file is basically the same as the flexspi xip kconfig file except that the compatible we are working with is just different. I wonder if there could be a common file instead like Kconfig.spi_xip, which can just use the same code but set the relevant compatible differently ? then we don't need to make similar changes in two places. If you think this would be too complex and don't want to attempt this then it's fine but I think there might be a missed opportunity here

lucien-nxp replied:
Yes, I agree your point. However, I think I should do this later than this PR merged, and then inform hake to test all the platforms affected by this change.

mmahadevan108 replied:
I agree with Declan. This extra file does not seem needed and we should be able to add the required changes to the cuurent Kconfig file. We need to add a tracking JIRA issue so we don't miss deleting this file if this is not being addressed as part of this PR.

decsny commented on soc/nxp/imxrt/soc.yml:
```diff
@@ -38,6 +38,14 @@ family:
+    - name: mimxrt798s
+      cpuclusters:
+      - name: cm33_cpu0
+      - name: cm33_cpu1
+      - name: hifi1
+      - name: hifi4
```
@nordicjm @tejlmand if these clusters not yet supported should they be in the soc.yml ?

nordicjm replied:
Would make sense to add them when support is added to keep things properly bisectable and minimal.

lucien-nxp replied:
So I need to remove hifi1 and hifi4 core provisionally, right? and add back once hifi core supported?

nordicjm replied:
Yes, doesn't make sense to be able to select a core in the build system that is impossible to use but will be listed as supported for a given board

dleach02 replied:
@lucien-nxp, please remove the cores that are not enabled yet on this platform.

lucien-nxp replied:
I have updated the yml. Sorry for the late respond. Originally, I wanted to wait for other comments to submit the changes together, so as to avoid multiple submission records.

dleach02 commented:
@nordicjm can you summarize what is needed to clear your change request? I'm having a hard time going through the comments in this PR to identify what @lucien-nxp needs to do to wrap this up.

nordicjm commented:
> @nordicjm can you summarize what is needed to clear your change request? I'm having a hard time going through the comments in this PR to identify what @lucien-nxp needs to do to wrap this up.

This outstanding part here has been waiting for an update: https://github.com/zephyrproject-rtos/zephyr/pull/79376/commits/5ac9605b449e9976a877eea23b4c519a72bce96f#r1894160349

lucien-nxp commented:
> > @nordicjm can you summarize what is needed to clear your change request? I'm having a hard time going through the comments in this PR to identify what @lucien-nxp needs to do to wrap this up.
> 
> This outstanding part here has been waiting for an update: [#79376 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/79376#discussion_r1894160349)

Thank you for pointing out. Could you help recheck and give me a approvement if no problem?

lucien-nxp commented:
#83873 , issues have been resolved.  Rebase to the newest commit node and do CI test again.

hakehuang commented:
board testing pass. only 3 cases with MPU region size limitation cases fail.

<html>
<body>

tests/kernel/sched/schedule_api/kernel.scheduler.multiq_no_timeslicing | v4.0.0-3243-gd8054384cfd5 | failed | Timeout | mimxrt700_evk/mimxrt798s/cm33_cpu0 | tests/kernel/sched/schedule_api
-- | -- | -- | -- | -- | --
tests/kernel/mem_heap/k_heap_api/kernel.k_heap_api | v4.0.0-3243-gd8054384cfd5 | error | Device issue (Flash error?) | mimxrt700_evk/mimxrt798s/cm33_cpu0 | tests/kernel/mem_heap/k_heap_api
tests/kernel/mem_protect/userspace/kernel.memory_protection.userspace | v4.0.0-3243-gd8054384cfd5 | failed | Failed harness:'Testsuite failed' | mimxrt700_evk/mimxrt798s/cm33_cpu0 | tests/kernel/mem_protect/userspace

</body>
</html>

dleach02 commented:
@hakehuang be careful removing DNM. This is set automatically due to the HAL side PR needing to be merged.

dleach02 commented:
@nordicjm @iuliana-prodan  @kartben @hakehuang need another round of +1 after merging the HAL and updating the west.yml
